src/window.rs

```diff
@@ -1346,6 +1347,18 @@ impl Window {
             .maybe_queue_on_main(move |w| w.set_cursor_icon(cursor))
     }
 
+    #[inline]
+    pub fn register_custom_cursor_icon(&self, key: u64, cursor: CustomCursor) {
+        self.window
+            .maybe_queue_on_main(move |w| w.register_custom_cursor_icon(key, cursor))
+    }
+
+    #[inline]
+    pub fn set_custom_cursor_icon(&self, key: u64) {
+        self.window
+            .maybe_queue_on_main(move |w| w.set_custom_cursor_icon(key))
+    }
```

Comment:
If I understand correctly API 2 didn't work out well because of the following:
- `CustomCursor` needs to know what backend it would be targeting for Linux (X11 or Wayland).
- Wayland needs access to the connection.
- Additionally now Web can't even create a `CustomCursor` without a `Window`. (potential solution discussed in https://github.com/rust-windowing/winit/pull/3218#discussion_r1390013504)

So I was thinking: could we just move `CustomCursor` creation to be a method on `Window`? E.g. `Window::create_custom_cursor(...) -> Result<CustomCursor>`, that way we can always determine the backend and Wayland can get all the access it needs. Also I was wondering if we could move this up to `EventLoopWindowTarget` instead of `Window`. Seems to me that a `CustomCursor` should work on any `Window`, at least on Web that would be nice.

That should do away with the ID system. WDYT?

---

Please don't do any changes yet, we need some consensus first before you do any unnecessary work.

Reply:
Yeah the current thing uses `maybe_queue_on_main` because I just naively used the same kind of code that was around. It would be a lot nicer to be able to actually return things and that could solve needing clunky ids. Though would threading issues persist when we aren't using this queue?

Reply:
Also dropping wayland cursors seems pretty hard if we don't have a global for the connection.

Reply:
It would be nice to reuse cursors between windows, so I have no issues with moving it to the `EventLoopWindowTarget`. I don't know that much about winit architecture but maybe that would work nicely in practice too.

Reply:
> Yeah the current thing uses `maybe_queue_on_main` because I just naively used the same kind of code that was around. It would be a lot nicer to be able to actually return things and that could solve needing clunky ids.

I think you are just looking for `maybe_wait_on_main()`, which is like `maybe_queue_on_main()`, but can actually return stuff.

> Also dropping wayland cursors seems pretty hard if we don't have a global for the connection.

Just spitballing here, but I was thinking that we could do some fun trickery with `Weak`, on `Drop` we can then turn that `Weak` into whatever it's holding and then properly deallocate that. That way if the event loop is destroyed before the user drops the cursor we can still deallocate internally.

In any case, we will probably have to evaluate some different ideas here and do some experiments.
Let me know what you think!

> It would be nice to reuse cursors between windows, so I have no issues with moving it to the `EventLoopWindowTarget`. I don't know that much about winit architecture but maybe that would work nicely in practice too.

From Winit's perspective this should work, unless it's somehow connected to specific windows on some backend. `EventLoopWindowTarget` is also not `Send + Sync`, so that might simplify some things.

Reply:
> I think you are just looking for `maybe_wait_on_main()`, which is like `maybe_queue_on_main()`, but can actually return stuff.

Seems to be what I needed from the start.

> Just spitballing here: but I was thinking that we could do some fun trickery with `Weak`, on `Drop` we can then turn that `Weak` into whatever it's holding and then properly deallocate that. That way if the event loop is destroyed before the user drops the cursor we can still deallocate internally.

I don't understand how that helps, you need to explain more. If we keep `Weak`s to cursors internally, then we can deallocate when event loop is dropped, but that would leave invalid cursors around. Also wouldn't help the problem of deallocating cursors dropped by the user when the event loops still exists.

> In any case, we will probably have to evaluate some different ideas here and do some experiments.
Let me know what you think!

A handle system could also work. In bevy, handles send drop events through a channel so the underlying storage knows to delete stuff ([StrongHandle](https://github.com/bevyengine/bevy/blob/bc9e159b2659435c4ca39eebd9b97396486161c2/crates/bevy_asset/src/handle.rs#L94-L101C2)). Though might be too annoying if its only really needed for wayland.

Reply:
> Edit: Thought I guess if we give out `Weak`s, keep `Arc`s, and require destruction through a method on the event loop, things could work out, though inconveniently.

That is what I meant! It's the other way around you interpreted it :)

> A handle system could also work. In bevy, handles send drop events through a channel so the underlying storage knows to delete stuff ([StrongHandle](https://github.com/bevyengine/bevy/blob/bc9e159b2659435c4ca39eebd9b97396486161c2/crates/bevy_asset/src/handle.rs#L94-L101C2)). Though might be too annoying if its only really needed for wayland.

That is indeed the only second solution I can come up as well. In any case, as discussed below, we should move that to a follow-up.